pkg/target/host.go

```diff
@@ -127,18 +114,22 @@ func (h *Host) Stream(wg *sync.WaitGroup) {
 				zap.Uint16("port", h.Port),
 				zap.Error(err),
 			)
-			err = h.Conn.Close()
-			if err != nil {
-				// not retrying here, file descriptor may be lost
-				h.Lg.Error("error closing the connection",
-					zap.Error(err))
-			}
-
-			h.Conn = nil
+			h.closeUpdateHostConnection(updateHostHealthStatus)
 		}
 	}
 }
 
+func (h *Host) closeUpdateHostConnection(updateHostHealthStatus chan *HostStatus) {
+	err := h.closeConnection()
+	if err != nil {
+		// not retrying here, file descriptor may be lost
+		h.Lg.Error("error closing the connection",
+			zap.Error(err))
+	}
+	h.updateHostConnection(nil)
+	h.updateHostHealthStatus(updateHostHealthStatus, false)
+}
+
 // Write does the remote write, i.e. sends the data
 func (h *Host) Write(b []byte) (nn int, err error) {
```

Comment:
Writing to the buffer can call Flush() when the buffer is full. Flush() will use the connection. 
We need to use the connection lock here.

Reply:
Also, I think the writer and connection locking is so interconnected that it probably makes sense to make a single RW mutex both for connection and writer. We can then lock it for writing when using the writer.